File: boj_codes/C++17/9663.cpp
// https://www.acmicpc.net/problem/9663
// First Written : 20210405
// Last Modified : 20210405

#include <cstdio>
#include <vector>

int v_visits[15] = {0};
int ld_visits[30] = {0};
int rd_visits[30] = {0};

std::vector<int> next_candidates(int depth, const int N){
    std::vector<int> candidates;
    for(int i = 0; i < N; i++){
        if(v_visits[i] == 1)
            continue;
        else if(ld_visits[i - depth + N - 1] == 1)
            continue;
        else if(rd_visits[i + depth] == 1)
            continue;

        candidates.push_back(i);
    }
    return candidates;
}

int backtrack(int depth, int total, const int N){
    if(depth == N){
        return total + 1;
    }

    else{
        std::vector<int> candidates = next_candidates(depth, N);

        for(auto j: candidates){
            v_visits[j] = 1;
            ld_visits[(j - depth + N - 1)] = 1;
            rd_visits[j + depth] = 1;

            total = backtrack(depth + 1, total, N);

            v_visits[j] = 0;
            ld_visits[(j - depth + N - 1)] = 0;
            rd_visits[j + depth] = 0;
            
        }

        return total;
    }
}

int main(){
    int N;
    scanf("%d", &N); 

    printf("%d", backtrack(0, 0, N));

    return 0;
}

File: boj_codes/Python3/PyPy3/9663.py
# https://www.acmicpc.net/problem/9663
# 20201101

import sys

N = int(sys.stdin.readline().rstrip())
v_visits = [0] * N
ld_visits = [0] * (2 * N - 1)
rd_visits = [0] * (2 * N - 1)


def next_candidates(depth):
    candidates = []

    for j in range(N):
        if v_visits[j] == 1:
            continue
        if ld_visits[j - depth + N - 1] == 1:
            continue
        if rd_visits[j + depth] == 1:
            continue

        candidates.append(j)

    return candidates


def backtrack(depth, total):
    if depth == N:
        return total + 1

    else:
        candidates = next_candidates(depth)

        for j in candidates:
            v_visits[j] = 1
            ld_visits[(j - depth + N - 1)] = 1
            rd_visits[j + depth] = 1

            total = backtrack(depth + 1, total)

            v_visits[j] = 0
            ld_visits[(j - depth + N - 1)] = 0
            rd_visits[j + depth] = 0

        del candidates
        return total


print(backtrack(0, 0))